pkg/vault/vault.go

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) WarnerMedia Direct, LLC. All rights reserved. Licensed under the MIT license.
+// See the LICENSE file for license information.
+
+/*
+Package vault provides methods for asserting that objects in Vault exist and match a given specification.
+
+Basics
+
+Methods in the vault package make use of the functional options pattern as a way to easily evolve interfaces
+without breaking backwards compatibility. Many methods require the use of the Vault LogicalClient struct,
+which is the object used to do most read / write / non-admin functions in the Go Vault client. For these,
+you can use the `WithLogicalClient` functional option.
+
+Examples
+
+Execute a method using a Vault client configured using standard Vault environment variables
+	import (
+		"os"
+		"testing"
+		"github.com/hbocodelabs/infratest/pkg/vault"
+	)
+
+	func TestVaultThing(t *testing.T) {
+		rootToken = os.GetEnv("VAULT_TOKEN")
+		vaultAddress = os.GetEnv("VAULT_ADDR")
+		ctx := context.Background()
+
+		clientConfig := &api.Config{
+			Address:    vaultAddress,
+			MaxRetries: 100,
+		}
+		client, err := api.NewClient(clientConfig)
+		require.Nil(t, err, "Vault NewClient method returned an unexpected error.")
+		client.SetToken(rootToken)
+
+		logicalClient := client.Logical()
+
+		expectedPath := "secret/data/hello"
+		expectedSecretData := map[string]interface{}{
+			"data": map[string]interface{}{
+				"username": "myname",
+				"password": "password",
+			},
+		}
+
+		// Do something that is supposed to create the expected secret here
+
+		vault.AssertVaultSecretExists(t, ctx, vault.WithLogicalClient(logicalClient), vault.WithPath(expectedPath), vault.WithKey("username"), vault.WithValue("myname"))
+		vault.AssertVaultSecretExists(t, ctx, vault.WithLogicalClient(logicalClient), vault.WithPath(expectedPath), vault.WithKey("password"), vault.WithValue("password"))
+	}
+*/
+package vault
+
+import (
+	// "context"
+	// "testing"
+
+	"context"
+
+	"github.com/hashicorp/vault/api"
+	"github.com/hbocodelabs/infratest/pkg/test"
+	"github.com/stretchr/testify/require"
+)
+
+//go:generate mockgen -destination=../../mock/vault.go -package=mock github.com/hbocodelabs/infratest/pkg/vault VaultLogicalClient
+
+const (
+	const_failure_missing_key       = "Secret did not contain the desired key."
+	const_failure_key_not_specified = "A Key must be set by one or more passed functional options."
+	const_failure_value_not_match   = "The specified value does not match the actual value."
```

Comment:
Nit: For whatever it's worth, common Golang linters (including the one used in my editor), dislike these `snake_case`'d constants and complain: `don't use underscores in Go names; const const_failure_missing_key should be constFailureMissingKey`.

Reply:
Fixed!